### PR TITLE
feat(cluster-call): support confirm success after all mfa apply ok

### PR DESCRIPTION
**Pull request by zhongwencool**
https://github.com/emqx/eip/pull/43
- support return result after all nodes MFA apply ok.
- support skip one failed commit by a specific node.
- support alarm when mfa apply failed.

### CODE REVIEW

**Comment by zhongwencool** [comment]
@zmstone ping

File: apps/emqx_gateway/src/lwm2m/lwm2m_xml/LWM2M_Server-v1_0.xml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (zmstone) [inline apps/emqx_gateway/src/lwm2m/lwm2m_xml/LWM2M_Server-v1_0.xml:1]: why are these files changed ?

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 2]
```diff
@@ -275,24 +292,91 @@ trans_query(TnxId) ->
...
+    end;
+confirm_commit({ok, TnxId, _Res} = Init, SucceedNum, Gap, Timeout) ->
+    timer:sleep(Gap),
+    case node_already_commit(TnxId, SucceedNum) of
+        ok -> Init;
+        _Error -> confirm_commit(Init, SucceedNum, Gap, Timeout - Gap)
+    end;
+confirm_commit(Error, _SucceedNum, _Gap, _Timeout) -> Error.
+
+node_not_commit(TnxId) ->
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:363]: maybe name it `lagging_nodes` ?

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 3]
```diff
@@ -275,24 +292,91 @@ trans_query(TnxId) ->
...
+    case transaction(fun commit_status_trans/2, ['<', TnxId]) of
+        {atomic, []} -> ok;
+        {atomic, Nodes} -> {error, Nodes}
+    end.
+
+node_already_commit(TnxId, SucceedNum) ->
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:369]: maybe name it `synced_nodes` ?

File: apps/emqx_machine/src/emqx_cluster_rpc.erl (hunk 2 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_cluster_rpc.erl:363]: and return `[]` to indicate none of the nodes are lagging

File: apps/emqx_machine/src/emqx_cluster_rpc.erl (hunk 3 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_cluster_rpc.erl:369]: and return `all` to indicate all nodes are synced

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 4]
```diff
@@ -275,24 +292,91 @@ trans_query(TnxId) ->
...
-        end,
-        {Succeed, Res}
-    catch
-        C : E ->
-            ?SLOG(critical, #{msg => "crash to apply MFA", tnx_id => TnxId, mfa => MFA, exception => C, reason => E}),
-            {false, lists:flatten(io_lib:format("TnxId(~p) apply MFA(~p) crash", [TnxId, MFA]))}
+    Result =
+        try
+            Meta0 = #{tnx_id => TnxId, module => M, function => F, args => ?TO_BIN(A)},
+            Res = erlang:apply(M, F, A),
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:301]: `Res = try erlang:apply(M, F, A), catch C : E : St -> {error, {C, E, St}} end,`

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 5]
```diff
@@ -76,28 +75,35 @@ start_link(Node, Name, RetryMs) ->
...
+                %% don't need rpc again inside transaction.
+                case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
+                    {ok, Node} -> gen_server:call({?MODULE, Node}, MFA, Timeout);
+                    disconnected -> {error, disconnected}
+                end
+        end,
+    End = erlang:monotonic_time(),
+    Gap = erlang:convert_time_unit(Begin - End, native, millisecond) + 50,
+    RetryMs = get_retry_ms(),
+    confirm_commit(InitRes, SucceedNum, Gap, 3 * max(Gap, RetryMs)).
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:106]: maybe name it: `wait_for_all_nodes_commit`

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 6]
```diff
@@ -76,28 +75,35 @@ start_link(Node, Name, RetryMs) ->
...
+                %% the initiate transaction must happened on core node
+                %% make sure MFA(in the transaction) and the transaction on the same node
+                %% don't need rpc again inside transaction.
+                case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
+                    {ok, Node} -> gen_server:call({?MODULE, Node}, MFA, Timeout);
+                    disconnected -> {error, disconnected}
+                end
+        end,
+    End = erlang:monotonic_time(),
+    Gap = erlang:convert_time_unit(Begin - End, native, millisecond) + 50,
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:104]: `Gap` is perhaps not the perfect name? maybe `MinDelay`

File: apps/emqx_machine/src/emqx_cluster_rpc.erl (hunk 5 above)
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:106]: Maybe calculate a `Deadline` before starting the wait loop, instead of calculating the next `Timeout` while looping.

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 7]
```diff
@@ -275,24 +292,91 @@ trans_query(TnxId) ->
...
+                Error -> Error
+            end
+    end;
+confirm_commit({ok, TnxId, _Res} = Init, SucceedNum, Gap, Timeout) ->
+    timer:sleep(Gap),
+    case node_already_commit(TnxId, SucceedNum) of
+        ok -> Init;
+        _Error -> confirm_commit(Init, SucceedNum, Gap, Timeout - Gap)
+    end;
+confirm_commit(Error, _SucceedNum, _Gap, _Timeout) -> Error.
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:361]: these function clauses are rather convoluted, please refactor to make it more readable

File: apps/emqx_gateway/src/lwm2m/lwm2m_xml/LWM2M_Server-v1_0.xml (hunk 1 above)
Comment (zhongwencool) [reply apps/emqx_gateway/src/lwm2m/lwm2m_xml/LWM2M_Server-v1_0.xml:1]: I run `find . -type f -print0 | xargs -0 dos2unix` to convert dos to unix format.
the command causes the change.

if I run `make clean-all & make ` on my mac machine will cause a lot files report:
`Bash script and /bin/bash^M: bad interpreter: No such file or directory`.
There are some files that are c library dependent.
so I run dos2unix command to covert this.

**Comment by zhongwencool** [comment]
@zmstone ping again :0

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 8]
```diff
@@ -76,27 +75,48 @@ start_link(Node, Name, RetryMs) ->
...
+-spec multicall(Module, Function, Args, SucceedNum, Timeout) -> {ok, TnxId, term()} |{error, Reason} when
     Module :: module(),
     Function :: atom(),
     Args :: [term()],
+    SucceedNum :: pos_integer() | all,
     TnxId :: pos_integer(),
     Timeout :: timeout(),
     Reason :: string().
-multicall(M, F, A, Timeout) ->
+multicall(M, F, A, RequireNum, Timeout)when RequireNum >= 1 ->
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:88]: missing a space after )

Comment (zmstone) [reply apps/emqx_machine/src/emqx_cluster_rpc.erl:88]: this guard is relying on the fact that `atom() > integer()`?
maybe check them separately .

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 9]
```diff
@@ -76,27 +75,48 @@ start_link(Node, Name, RetryMs) ->
...
+                %% make sure MFA(in the transaction) and the transaction on the same node
+                %% don't need rpc again inside transaction.
+                case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
+                    {ok, Node} -> gen_server:call({?MODULE, Node}, MFA, Timeout);
+                    disconnected -> {error, disconnected}
+                end
+        end,
+    End = erlang:monotonic_time(),
+    MinDelay = erlang:convert_time_unit(Begin - End, native, millisecond) + 50,
+    %% Failed after 3 attempts.
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:105]:
```suggestion
    %% Fail after 3 attempts.
```

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 10]
```diff
@@ -275,24 +306,71 @@ trans_query(TnxId) ->
...
+        false when Remain > 0 ->
+            wait_for_nodes_commit(RequiredNum, TnxId, Delay, Remain - Delay);
+        false ->
+            case legging_node(TnxId) of
+                [] -> ok; %% All commit but The succeedNum > length(nodes()).
+                Nodes -> {error, Nodes}
+            end
+    end.
+
+legging_node(TnxId) ->
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:361]: typo: lagging

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 11]
```diff
@@ -275,24 +306,71 @@ trans_query(TnxId) ->
...
-            _ ->
-                ?SLOG(error, #{msg => "failed to apply MFA", tnx_id => TnxId, mfa => MFA, result => Res}),
-                false
+-define(TO_BIN(_B_), iolist_to_binary(io_lib:format("~p", [_B_]))).
+
+apply_mfa(TnxId, {M, F, A}) ->
+    Res =
+        try erlang:apply(M, F, A)
+        catch
+            C : E -> {crash, C, E}
```
[the remark this answers is not in the document]
Comment (zmstone) [reply apps/emqx_machine/src/emqx_cluster_rpc.erl:315]: maybe return `{error, #{exception => C, reason => E, stacktrace => St}}`,
then there is no need for a special clause to handle the `crash` tuple.

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 12]
```diff
@@ -275,24 +306,71 @@ trans_query(TnxId) ->
...
         end,
-        {Succeed, Res}
-    catch
-        C : E ->
-            ?SLOG(critical, #{msg => "crash to apply MFA", tnx_id => TnxId, mfa => MFA, exception => C, reason => E}),
-            {false, lists:flatten(io_lib:format("TnxId(~p) apply MFA(~p) crash", [TnxId, MFA]))}
+    Meta = #{tnx_id => TnxId, module => M, function => F, args => ?TO_BIN(A)},
+    log_and_alarm(Res, Meta),
+    Succeed = (Res =:= ok orelse (is_tuple(Res) andalso Res =/= {} andalso element(1, Res) =:= ok)),
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:319]: the `ok` or `{ok, _}` check are duplicated in these two lines.
maybe create a `is_success/1` help function, to check the result, then call log_and_alarm before returning the result.

**Review by zmstone — COMMENTED** [review]
sorry for the delay.

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 13]
```diff
@@ -275,24 +306,71 @@ trans_query(TnxId) ->
...
-             ok ->
-                 ?SLOG(notice, #{msg => "succeeded to apply MFA", tnx_id => TnxId, mfa => MFA, result => Res}),
-                 true;
-            {ok, _} ->
-                ?SLOG(notice, #{msg => "succeeded to apply MFA", tnx_id => TnxId, mfa => MFA, result => Res}),
-                true;
-            _ ->
-                ?SLOG(error, #{msg => "failed to apply MFA", tnx_id => TnxId, mfa => MFA, result => Res}),
-                false
+-define(TO_BIN(_B_), iolist_to_binary(io_lib:format("~p", [_B_]))).
```
[the remark this answers is not in the document]
Comment (zhongwencool) [reply apps/emqx_machine/src/emqx_cluster_rpc.erl:309]: This meta will also be used by alarm, so it needs to be formated.

File: apps/emqx_machine/src/emqx_cluster_rpc.erl [hunk 14]
```diff
@@ -275,24 +306,68 @@ trans_query(TnxId) ->
...
+        false when Remain > 0 ->
+            wait_for_nodes_commit(RequiredNum, TnxId, Delay, Remain - Delay);
+        false ->
+            case lagging_node(TnxId) of
+                [] -> ok; %% All commit but The succeedNum > length(nodes()).
+                Nodes -> {error, Nodes}
+            end
+    end.
+
+lagging_node(TnxId) ->
```
Comment (zmstone) [inline apps/emqx_machine/src/emqx_cluster_rpc.erl:358]:
```suggestion
lagging_nodes(TnxId) ->
```
```suggestion
lagging_node(TnxId) ->
```

**Review by zmstone — DISMISSED** [review]

**Review by zmstone — APPROVED** [review]
